src/Analyser/NodeScopeResolver.php

```diff
@@ -1561,23 +1561,35 @@ private function ensureShallowNonNullability(MutatingScope $scope, Expr $exprToS
 	{
 		$exprType = $scope->getType($exprToSpecify);
 		$exprTypeWithoutNull = TypeCombinator::removeNull($exprType);
+		$specifiedExpressions = [];
 		if (!$exprType->equals($exprTypeWithoutNull)) {
 			$nativeType = $scope->getNativeType($exprToSpecify);
 			$scope = $scope->specifyExpressionType(
 				$exprToSpecify,
 				$exprTypeWithoutNull,
 				TypeCombinator::removeNull($nativeType),
 			);
+			$specifiedExpressions[] = new EnsuredNonNullabilityResultExpression($exprToSpecify, $exprType, $nativeType);
+		}
 
-			return new EnsuredNonNullabilityResult(
-				$scope,
-				[
-					new EnsuredNonNullabilityResultExpression($exprToSpecify, $exprType, $nativeType),
-				],
-			);
+		if ($exprToSpecify instanceof ArrayDimFetch && $exprToSpecify->dim !== null) {
+			$dimFetchVarType = $scope->getType($exprToSpecify->var);
+			if ($dimFetchVarType instanceof ConstantArrayType) {
```

Comment:
I worry this logic isn't really sufficient. You can also have a union of constant arrays. And I guess it should also do something aboout nested ArrayDimFetch - `$a['a']['b']['c']` where the intermediate keys are also optional...

I feel like instead of reimplementing the logic here, this place should call TypeSpecifier because the place where it handles `Isset_` looks like it has this covered: https://github.com/phpstan/phpstan-src/blob/5c48a180115306c5ca005c2617d8b27db1c2df49/src/Analyser/TypeSpecifier.php#L742-L826

So the scope after `$x ?? throw` should be something like `$scope->filterByTruthyValue(new Isset_($expr->left))` (which delegates the call to TypeSpecifier internally).